blog/content/post/2019-05-29-spectre-and-meltdown-on-apu2.md

```diff
@@ -0,0 +1,360 @@
+Reading at malicious_x = 0xffffffffffdfeedc... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeedd... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeede... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeedf... Success: 0xFF=’?’ score=0 
+```
+
+> One can see that the secret string was not disclosed (0xFF=? everywhere).
+> Let's increase the cache hit threshold.
+
+```
+./spectre.out 70
+Version: commit 856f80f2937f2bb812cab68d45c149272a1783d5
+Using a cache hit threshold of 70.
+Build: RDTSCP_SUPPORTED MFENCE_SUPPORTED CLFLUSH_SUPPORTED INTEL_MITIGATION_DISABLED LINUX_KERNEL_MITIGATION_DISABLED 
+Reading 40 bytes:
+Reading at malicious_x = 0xffffffffffdfeeb8... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeeb9... Success: 0x68=’h’ score=2 
+Reading at malicious_x = 0xffffffffffdfeeba... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeebb... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeebc... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeebd... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeebe... Success: 0x67=’g’ score=2 
+...
+Reading at malicious_x = 0xffffffffffdfeed9... Success: 0x69=’i’ score=2 
+Reading at malicious_x = 0xffffffffffdfeeda... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeedb... Success: 0x72=’r’ score=2 
+Reading at malicious_x = 0xffffffffffdfeedc... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeedd... Success: 0x67=’g’ score=2 
+Reading at malicious_x = 0xffffffffffdfeede... Success: 0xFF=’?’ score=0 
+Reading at malicious_x = 0xffffffffffdfeedf... Success: 0xFF=’?’ score=0 
+```
+
+One can see that positive results are obtained with a larger value of cache hit
+threshold (with 70 some characters have not been obtained - score 0, 100 is sufficient
+to get whole string - score 2 everywhere). It is related to CPU performance.
+
+> For those more interested in details of the attack, please refer to
+> [Spectre Attacks](https://spectreattack.com/spectre.pdf) or to my
+> [previous post](https://blog.3mdeb.com/2019/2019-03-14-meltdown-and-spectre-vulnerabilities/)
+> introducing the Meltdown and Spectre.
+
+Let's also check out the kernel mitigation. Rebuild the PoC with:
+
+```shell
+make clean
+CFLAGS=-DLINUX_KERNEL_MITIGATION make
+```
+
+After trying to run the PoC with 20/70/100 the string is not disclosed. But
+what if the kernel we use does not have the mitigation? The answer is firmware
+and microcode for full hardware mitigation.
+
+# Microcode update and spectre
+
+To present results of Spectre vulnerability without kernel mitigation, we will
+need the microcode update. For details how to build PC Engines apu2 coreboot
+firmware with microcode update, please refer to this [guide](https://github.com/pcengines/apu2-documentation/blob/master/docs/microcode_patching.md)
```

Comment:
@miczyg1 does this procedure still work fine? There are some bug reports related: https://github.com/pcengines/apu2-documentation/issues/145

Reply:
@pietrushnic this bug is related to docker itself not to build process. I have an asciinema in the pst showing how to build the image. Worked perfectly fine for the version mentioned in blogpost.

Reply:
Also checked with new v4.9.0.5. No problems with building and flashing. For mainline releases we use unmodified coreboot-sdk container, so it is the Linxu configuration problem with permissions. It is highly dependable on groups, directory configuration, permissions, structure. May even be a file system (like if someone would like to compile coreboot on NTFS for some reason). Anyway, I have built is successfully without problems, waiting for the response to my comment in linked issue.

Reply:
Ok